### PR TITLE
Fix 'config_section' and add ability to parse multiple sections

### PR DESCRIPTION
**Pull request by manchoz**
- Fix passing `config_section` with arbitrary name not working anymore.
- Read parameters from multiple-sections config-file with `config_section` as a list of sections names.

eg.
```python
@begin.start(config_file='conf.ini', config_section=['foo', 'bar'])
def main(value1=None, value2=None):
    print('value1: {}, value2:{}'.format(value1, value2))
```

with `conf.ini` as

```ini
[foo]
value1 = asd

[bar]
value2 = 123
```

### CODE REVIEW

File: begin/cmdline.py [hunk 1]
```diff
@@ -58,10 +58,16 @@ def from_name(self, name, default=NODEFAULT, section=None):
...
-            try:
-                default = self._parser.get(section, name)
-            except (configparser.NoSectionError, configparser.NoOptionError):
-                pass
+            if isinstance(section, list):
+                sections = section
+            else:
+                sections = []
+                sections.append(section)
+            for section in sections:
```
Comment (sh0oki) [inline begin/cmdline.py]: I'm not sure I agree with this behavior. 
We need to find a way to make this more deterministic in my opinion.
WDYT?

File: begin/cmdline.py [hunk 2]
```diff
@@ -58,10 +58,16 @@ def from_name(self, name, default=NODEFAULT, section=None):
         "Get default value from argument name"
         if len(self._parser.sections()) > 0:
             section = self._section if section is None else section
-            try:
-                default = self._parser.get(section, name)
-            except (configparser.NoSectionError, configparser.NoOptionError):
-                pass
+            if isinstance(section, list):
```
Comment (sh0oki) [inline begin/cmdline.py]: I'd test if `not isinstance(section, str)`, as users can pass a tuple or a set here, too.

File: begin/cmdline.py [hunk 3]
```diff
@@ -58,10 +58,16 @@ def from_name(self, name, default=NODEFAULT, section=None):
...
+                sections = section
+            else:
+                sections = []
+                sections.append(section)
+            for section in sections:
+                try:
+                    default = self._parser.get(section, name)
```
Comment (sh0oki) [inline begin/cmdline.py]: break if found

File: begin/cmdline.py [hunk 4]
```diff
@@ -163,7 +169,8 @@ def create_parser(func, env_prefix=None, config_file=None, config_section=None,
     arguments will raise a ValueError exception. A prefix on expected
     environment variables can be added using the env_prefix argument.
     """
-    defaults = DefaultsManager(env_prefix, config_file, func.__name__)
+    section = config_section if config_section is not None else func.__name__
```
Comment (sh0oki) [inline begin/cmdline.py]: I'm not sure I understand this change. Can you please explain?

**Review by sh0oki — CHANGES_REQUESTED** [review]
Hey @manchoz! Thanks for your contribution. 
I added a couple of comments, please have a look.
I'd also like to add some unit-tests to test the (1) fixed and (2) new behavior. 
Thanks!

File: begin/cmdline.py (hunk 2 above)
Comment (manchoz) [reply begin/cmdline.py]: Ok.

Would we like to permit users also to pass tuple, set, or string with space-separated sections names?

File: begin/cmdline.py (hunk 1 above)
Comment (manchoz) [reply begin/cmdline.py]: Ok.

Would be the case to assume that `DefaultsManager._section` is always a list and move all the type-checking logic to `DefaultsManager.set_config_section()`?

Also `DefaultsManager.set_config_section()` could be refactored to be a property setter for `DefaultsManager.section`. WDYT?

File: begin/cmdline.py (hunk 3 above)
Comment (manchoz) [reply begin/cmdline.py]: Ok.

File: begin/cmdline.py (hunk 4 above)
Comment (manchoz) [reply begin/cmdline.py]: That's the single point I could find which leads to setting `self._section` to the value passed via `config_section` in the `begin.start()` call. Otherwise `self._section` will always be set to `func.__name__` no matter what is passed in `begin.start()`. Am I missing something?

File: begin/cmdline.py [hunk 5]
```diff
@@ -37,10 +37,11 @@ def __init__(self, env_prefix=None, config_file=None, config_section=None):
         self._use_env = env_prefix is not None
         self._prefix = '' if not self._use_env else env_prefix
         self._parser = configparser.ConfigParser()
-        self._section = config_section
+        self._section = None
```
Comment (sh0oki) [inline begin/cmdline.py:40]: This is rather confusing (having both a `section`, `_section` and a property with this name.
Let's try to find a clearer name for this? Can we maybe delete some of these?

File: begin/cmdline.py [hunk 6]
```diff
@@ -56,18 +57,44 @@ def from_param(self, param, default=NODEFAULT):
...
-            section = self._section if section is None else section
-            try:
-                default = self._parser.get(section, name)
-            except (configparser.NoSectionError, configparser.NoOptionError):
-                pass
+        sections = self._get_list_section(section) if section is not None else self.section
+        if len(sections) > 0:
+            for sec in self.section:
+                try:
+                    default = self._parser.get(sec, name)
```
Comment (sh0oki) [inline begin/cmdline.py]: break here?

File: begin/cmdline.py [hunk 7]
```diff
@@ -56,18 +57,44 @@ def from_param(self, param, default=NODEFAULT):
...
     def from_name(self, name, default=NODEFAULT, section=None):
         "Get default value from argument name"
-        if len(self._parser.sections()) > 0:
-            section = self._section if section is None else section
-            try:
-                default = self._parser.get(section, name)
-            except (configparser.NoSectionError, configparser.NoOptionError):
-                pass
+        sections = self._get_list_section(section) if section is not None else self.section
+        if len(sections) > 0:
```
Comment (sh0oki) [inline begin/cmdline.py]: Why is this needed? Won't the for loop just do nothing if there are no sections?

**Review by sh0oki — CHANGES_REQUESTED** [review]

File: begin/cmdline.py (hunk 5 above)
Comment (manchoz) [reply begin/cmdline.py:40]: Just trying to hide instance variables... ``DefaultsManager._section`` is for private instance variable, which should never be referenced directly. You should use ``DefaultsManager.section``, instead, which has getter and setter methods.

File: begin/cmdline.py (hunk 7 above)
Comment (manchoz) [reply begin/cmdline.py]: Yes, you're right.

File: begin/cmdline.py (hunk 6 above)
Comment (manchoz) [reply begin/cmdline.py]: Ok.

**Comment by sh0oki** [comment]
@manchoz I found this stackoverflow article that might help the Python 2.6 error: https://stackoverflow.com/questions/21034954/valueerror-zero-length-field-name-in-format-in-python2-6-6

**Comment by manchoz** [comment]
Hi @sh0oki, thank you so much for merging!
